### PR TITLE
Optimized dropout forward

### PR DESCRIPTION
**Pull request by unnonouno**
I optimized dropout foward.

### CODE REVIEW

File: chainer/functions/noise/dropout.py [hunk 1]
```diff
@@ -25,10 +25,18 @@ def forward(self, x):
...
+                self.mask, y = cuda.elementwise(
+                    'T x, R r, T scale, T ratio', 'T mask, T y',
+                    '''
+                    mask = (r >= ratio) * scale;
+                    y = x * mask;
+                    ''',
+                    'dropout_fwd',
+                )(x[0], rand, scale, self.dropout_ratio)
+                return y,
         return x[0] * self.mask,
```
Comment (niboshi) [inline chainer/functions/noise/dropout.py]: Please put this `return` statement inside the `if` construct (i.e. just before `else` at line 29).

Comment (niboshi) [inline chainer/functions/noise/dropout.py]: `if` constructs and `return` statements are not balanced. It's hard to read.
How about this?

(I know `x[0] * self.mask` is repeated. But IMO it's better than skewed code.)


```
    def forward(self, x):
        self.retain_inputs(())
        y = None
        if hasattr(self, 'mask'):
            y = x[0] * self.mask
        else:
            scale = x[0].dtype.type(1. / (1 - self.dropout_ratio))
            xp = cuda.get_array_module(*x)
            if xp == numpy:
                flag = xp.random.rand(*x[0].shape) >= self.dropout_ratio
                self.mask = scale * flag
                y = x[0] * self.mask
            else:
                rand = xp.random.rand(*x[0].shape, dtype=numpy.float32)
                self.mask, y = cuda.elementwise(
                    'T x, R r, T scale, T ratio', 'T mask, T y',
                    '''
                    mask = (r >= ratio) * scale;
                    y = x * mask;
                    ''',
                    'dropout_fwd',
                )(x[0], rand, scale, self.dropout_ratio)
        return y,
```

**Comment by unnonouno** [comment]
It sounds good. I updated it.

**Comment by niboshi** [comment]
LGTM!

**Pull request merged**